gunrock/app/rw/rw_enactor.cuh

```diff
@@ -0,0 +1,385 @@
+#include <curand.h>
+#include <curand_kernel.h>
+// </DONE>
+
+
+namespace gunrock {
+namespace app {
+// <DONE> change namespace
+namespace rw {
+// </DONE>
+
+/**
+ * @brief Speciflying parameters for hello Enactor
+ * @param parameters The util::Parameter<...> structure holding all parameter info
+ * \return cudaError_t error message(s), if any
+ */
+cudaError_t UseParameters_enactor(util::Parameters &parameters)
+{
+    cudaError_t retval = cudaSuccess;
+    GUARD_CU(app::UseParameters_enactor(parameters));
+
+    // <DONE> if needed, add command line parameters used by the enactor here
+    // </DONE>
+    
+    return retval;
+}
+
+/**
+ * @brief defination of hello iteration loop
+ * @tparam EnactorT Type of enactor
+ */
+template <typename EnactorT>
+struct RWIterationLoop : public IterationLoopBase
+    <EnactorT, Use_FullQ | Push
+    // <OPEN>if needed, stack more option, e.g.:
+    // | (((EnactorT::Problem::FLAG & Mark_Predecessors) != 0) ?
+    // Update_Predecessors : 0x0)
+    // </OPEN>
+    >
+{
+    typedef typename EnactorT::VertexT VertexT;
+    typedef typename EnactorT::SizeT   SizeT;
+    typedef typename EnactorT::ValueT  ValueT;
+    typedef typename EnactorT::Problem::GraphT::CsrT CsrT;
+    typedef typename EnactorT::Problem::GraphT::GpT  GpT;
+    
+    typedef IterationLoopBase
+        <EnactorT, Use_FullQ | Push
+        // <OPEN> add the same options as in template parameters here, e.g.:
+        // | (((EnactorT::Problem::FLAG & Mark_Predecessors) != 0) ?
+        // Update_Predecessors : 0x0)
+        // </OPEN>
+        > BaseIterationLoop;
+
+    RWIterationLoop() : BaseIterationLoop() {}
+
+    /**
+     * @brief Core computation of hello, one iteration
+     * @param[in] peer_ Which GPU peers to work on, 0 means local
+     * \return cudaError_t error message(s), if any
+     */
+    cudaError_t Core(int peer_ = 0)
+    {
+        // --
+        // Alias variables
+        
+        auto &data_slice = this -> enactor ->
+            problem -> data_slices[this -> gpu_num][0];
+        
+        auto &enactor_slice = this -> enactor ->
+            enactor_slices[this -> gpu_num * this -> enactor -> num_gpus + peer_];
+        
+        auto &enactor_stats    = enactor_slice.enactor_stats;
+        auto &graph            = data_slice.sub_graph[0];
+        auto &frontier         = enactor_slice.frontier;
+        auto &oprtr_parameters = enactor_slice.oprtr_parameters;
+        auto &retval           = enactor_stats.retval;
+        auto &iteration        = enactor_stats.iteration;
+        
+        // <DONE> add problem specific data alias here:
+        auto &walks       = data_slice.walks;
+        auto &rand        = data_slice.rand;
+        auto &walk_length = data_slice.walk_length;
+        auto &gen         = data_slice.gen;
+        // </DONE>
+        
+        curandGenerateUniform(gen, rand.GetPointer(util::DEVICE), graph.nodes);
+        
+        GUARD_CU(frontier.V_Q()->ForAll(
```

Comment:
Noting that this doesn't do the sorted RandomWalk, which doesn't seem hard because you can use CUB's blocksort to implement that. I think my initial take on it just load balancing nicely might be an incorrect assumption. I'll see if I can find Shari's notes on that.

Reply:
OK sounds good -- I didn't look closely at the sorted code.  If you have any references for what Shari was trying to do w/ the sorting I'd love to take a look

Reply:
Array1D provides the sort() function https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/oprtr/1D_oprtr/sort.cuh#L35 here.

`ForAll` assumes all per-element workload to have ~the same cost, and it doesn't do load balancing. I don't think the random access could cost more than 3 sorting, but I maybe wrong here.

Reply:
I think it will be interesting to get the numbers with sort as well then, because per Shari's code, that performed better than random accesses.

Reply:
Sorting potentially increases locality of reference, which leads to better coalescing. It's a win if there is any locality. It is frankly hard to come up with principled ways to improve random walks because one would imagine you would be completely gated by the throughput of random memory accesses, and it's hard to extract any locality from randomness.